Source/WTF/wtf/threads/Signals.cpp

```diff
@@ -203,6 +204,162 @@ kern_return_t catch_mach_exception_raise_state_identity(mach_port_t, mach_port_t
+            continue;
+        }
+        mach_port_deallocate(mach_task_self(), threads[i]);
+    }
+    return thread;
+}
+
+kern_return_t catch_mach_exception_raise_state(
+    mach_port_t,
+    exception_type_t,
+    const mach_exception_data_t,
+    mach_msg_type_number_t,
+    int*,
+    const thread_state_t,
+    mach_msg_type_number_t,
+    thread_state_t,
+    mach_msg_type_number_t*)
+{
+    dataLogLn("We should not have called catch_mach_exception_raise_state, please file a bug at bugs.webkit.org");
+    return KERN_FAILURE;
+}
+
+kern_return_t catch_mach_exception_raise_identity_protected(
+    mach_port_t exceptionPort,
+    uint64_t threadID,
+    mach_port_t taskIDToken,
+    exception_type_t exceptionType,
+    mach_exception_data_t exceptionData,
+    mach_msg_type_number_t dataCount)
+{
+    UNUSED_PARAM(taskIDToken);
+
+    SignalHandlers& handlers = g_wtfConfig.signalHandlers;
+    RELEASE_ASSERT(exceptionPort == handlers.exceptionPort);
+    // If we wanted to distinguish between SIGBUS and SIGSEGV for EXC_BAD_ACCESS on Darwin we could do:
+    // if (exceptionData[0] == KERN_INVALID_ADDRESS)
+    //    signal = SIGSEGV;
+    // else
+    //    signal = SIGBUS;
+    Signal signal = fromMachException(exceptionType);
+    RELEASE_ASSERT(signal != Signal::Unknown);
+
+    thread_act_t thread = threadIDGetThreadPort(threadID);
+
+    mach_msg_type_number_t stateCount = 0;
+
+#if CPU(X86_64) || CPU(X86)
+    x86_thread_state_t state;
+    int flavor = x86_THREAD_STATE;
+    stateCount = x86_UNIFIED_THREAD_STATE_COUNT;
+#elif CPU(ARM64) || CPU(ARM)
+    arm_unified_thread_state state;
+    int flavor = ARM_THREAD_STATE;
+    stateCount = ARM_UNIFIED_THREAD_STATE_COUNT;
+#endif
+
+    kern_return_t kr = thread_get_state(thread, flavor, (thread_state_t)&state, &stateCount);
+    if (kr != KERN_SUCCESS) {
+        dataLogLn("thread_get_state failed due to ", mach_error_string(kr));
+        return kr;
+    }
+
+#if CPU(ARM64E) && OS(DARWIN)
+    ptrauth_generic_signature_t inStateHash = hashThreadState((thread_state_t)&state);
+#endif
+
+#if CPU(X86_64)
+    PlatformRegisters& registers = state.uts.ts64;
+    ASSERT(state.ash.flavor == X86_THREAD_STATE64);
+#elif CPU(X86)
+    PlatformRegisters& registers = state.uts.ts32;
+    ASSERT(state.ash.flavor == X86_THREAD_STATE32);
+#elif CPU(ARM64)
+    PlatformRegisters& registers = state.ts_64;
+    ASSERT(state.ash.flavor == ARM_THREAD_STATE64);
+#elif CPU(ARM)
+    PlatformRegisters& registers = state.ts_32;
+    ASSERT(state.ash.flavor == ARM_THREAD_STATE32);
+#endif
+
+    bool didHandle = false;
+    kr = runSignalHandlers(signal, registers, didHandle, dataCount, exceptionData);
+    if (kr != KERN_SUCCESS)
+        return kr;
+
+    if (didHandle) {
+#if CPU(ARM64E) && OS(DARWIN)
+        RELEASE_ASSERT(inStateHash == hashThreadState((thread_state_t)&state));
+#endif
+        kr = thread_set_state(thread, flavor, (thread_state_t)&state,
+            stateCount);
```

Comment:
Don't we also need to `mach_port_deallocate` `thread` here after this last use of it?